library/alloc/src/collections/btree/map.rs

```diff
@@ -2762,287 +2763,271 @@ pub struct CursorMut<
     ///
-    /// This returns `None` if the cursor is currently pointing to the
-    /// "ghost" non-element.
+    /// If the cursor is at the end of the map then `None` is returned
     #[unstable(feature = "btree_cursors", issue = "107540")]
-    pub fn key(&self) -> Option<&K> {
-        self.current.as_ref().map(|current| current.reborrow().into_kv().0)
+    pub fn peek_next(&mut self) -> Option<(&K, &mut V)> {
+        let (k, v) = self.inner.peek_next()?;
+        Some((&*k, v))
     }
 
-    /// Returns a reference to the value of the element that the cursor is
-    /// currently pointing to.
+    /// Returns a reference to the the key and value of the previous element
+    /// without moving the cursor.
     ///
-    /// This returns `None` if the cursor is currently pointing to the
-    /// "ghost" non-element.
+    /// If the cursor is at the start of the map then `None` is returned.
     #[unstable(feature = "btree_cursors", issue = "107540")]
-    pub fn value(&self) -> Option<&V> {
-        self.current.as_ref().map(|current| current.reborrow().into_kv().1)
+    pub fn peek_prev(&mut self) -> Option<(&K, &mut V)> {
+        let (k, v) = self.inner.peek_prev()?;
+        Some((&*k, v))
     }
 
-    /// Returns a reference to the key and value of the element that the cursor
-    /// is currently pointing to.
+    /// Returns a read-only cursor pointing to the same location as the
+    /// `CursorMut`.
     ///
-    /// This returns `None` if the cursor is currently pointing to the
-    /// "ghost" non-element.
+    /// The lifetime of the returned `Cursor` is bound to that of the
+    /// `CursorMut`, which means it cannot outlive the `CursorMut` and that the
+    /// `CursorMut` is frozen for the lifetime of the `Cursor`.
     #[unstable(feature = "btree_cursors", issue = "107540")]
-    pub fn key_value(&self) -> Option<(&K, &V)> {
-        self.current.as_ref().map(|current| current.reborrow().into_kv())
+    pub fn as_cursor(&self) -> Cursor<'_, K, V> {
+        self.inner.as_cursor()
     }
 
-    /// Returns a mutable reference to the value of the element that the cursor
-    /// is currently pointing to.
+    /// Converts the cursor into a [`CursorMutKey`], which allows mutating
+    /// the key of elements in the tree.
+    ///
+    /// # Safety
+    ///
+    /// Since this cursor allows mutating keys, you must ensure that the `BTreeMap`
+    /// invariants are maintained. Specifically:
     ///
-    /// This returns `None` if the cursor is currently pointing to the
-    /// "ghost" non-element.
+    /// * The key of the newly inserted element must be unique in the tree.
+    /// * All keys in the tree must remain in sorted order.
     #[unstable(feature = "btree_cursors", issue = "107540")]
-    pub fn value_mut(&mut self) -> Option<&mut V> {
-        self.current.as_mut().map(|current| current.kv_mut().1)
+    pub unsafe fn with_mutable_key(self) -> CursorMutKey<'a, K, V, A> {
+        self.inner
     }
```

Comment:
I'm skeptical that maintaining `BTreeMap` invariants is safety critical because, if it was, then the following safe code would be unsound:
```rust
struct Key;
impl Ord for Key {
    fn cmp(&self, other: &Self) -> Ordering {
        [Ordering::Less, Ordering::Equal, Ordering::Greater].choose(&mut thread_rng())
    }
}
```
And, from the docs, we have:
> It is a logic error for a key to be modified in such a way that the key’s ordering relative to any other key, as determined by the [Ord](https://doc.rust-lang.org/beta/std/cmp/trait.Ord.html) trait, changes while it is in the map...The behavior resulting from such a logic error is not specified, but will be encapsulated to the BTreeMap that observed the logic error and **not result in undefined behavior.**

Reply:
Second, [I don't see `mutable` being spelled out in any ident in `std`](https://doc.rust-lang.org/beta/std/index.html?search=mutable). It's always abbreviated to `mut`.

Reply:
We want that unsafe code can trust the order of `BTreeMap<usize, _>`. If you can get a &mut reference to a key you can swap it with a different value even if the type (like usize) is entirely well-behaved.

Reply:
Ok good to know! My point about the name still stands.

Reply:
As I mentioned in the tracking issue, I *much* prefer leveraging the entry API and adding key mutation there (see: #112896) rather than adding an entire other API like this.

Reply:
Agreed